File: frontend/src/types/blocks.ts
// Mirror of backend Pydantic schemas

export type BlockStatus = 'complete' | 'partial' | 'error'

export interface ReportBlockBase {
  block_id: string
  block_type: string
  title: string
  order: number
  status: BlockStatus
  warnings: string[]
  source_task_ids: string[]
  created_by_dept: string
  created_at: string
}

export interface TextBlock extends ReportBlockBase {
  block_type: 'text'
  content: string
  style: 'body' | 'heading' | 'callout' | 'quote'
  language: string
}

export interface KPICard extends ReportBlockBase {
  block_type: 'kpi_card'
  metric_name: string
  value: number
  unit: string
  delta?: number
  delta_pct?: number
  trend?: 'up' | 'down' | 'stable'
  is_positive_trend?: boolean
  benchmark?: number
  verification_status: 'verified' | 'unverified'
  source_reference: string
}

export interface TableColumn {
  key: string
  label: string
  dtype: 'number' | 'string' | 'date' | 'percent' | 'currency'
  unit?: string
  sortable: boolean
  highlight_rule?: Record<string, unknown>
}

export interface TableBlock extends ReportBlockBase {
  block_type: 'table'
  columns: TableColumn[]
  rows: Record<string, unknown>[]
  totals_row?: Record<string, unknown>
  pagination: boolean
  exportable: boolean
  source_file_id: string
  source_location: string
}

export interface ChartBlock extends ReportBlockBase {
  block_type: 'chart'
  chart_type: string
  vega_lite_spec: Record<string, unknown>
  caption?: string
  data_source_description: string
  interactive: boolean
}

export interface InsightBlock extends ReportBlockBase {
  block_type: 'insight'
  insight_type: 'finding' | 'anomaly' | 'recommendation' | 'risk' | 'opportunity'
  severity: 'info' | 'warning' | 'critical'
  headline: string
  explanation: string
  supporting_data: string[]
  confidence: number
}

export interface ComparisonItem {
  label: string
  metrics: Record<string, number | string>
}

export interface ComparisonBlock extends ReportBlockBase {
  block_type: 'comparison'
  items: ComparisonItem[]
  dimensions: string[]
  winner?: string
  analysis: string
}

export interface TimePoint {
  period: string
  value: number
}

export interface ForecastPoint {
  period: string
  value: number
  lower_bound: number
  upper_bound: number
}

export interface ForecastBlock extends ReportBlockBase {
  block_type: 'forecast'
  metric_name: string
  historical: TimePoint[]
  forecast: ForecastPoint[]
  method: string
  confidence_interval: number
  model_accuracy: Record<string, number>
}

export interface RiskItem {
  name: string
  probability: 'low' | 'medium' | 'high'
  impact: 'low' | 'medium' | 'high'
  description: string
  mitigation?: string
}

export interface RiskMatrixBlock extends ReportBlockBase {
  block_type: 'risk_matrix'
  risks: RiskItem[]
  overall_risk_level: 'low' | 'medium' | 'high' | 'critical'
}

export interface ExecutiveSummaryBlock extends ReportBlockBase {
  block_type: 'executive_summary'
  key_findings: string[]
  recommendations: string[]
  overall_conclusion: string
  report_quality_score: number
}

export type ReportBlock =
  | TextBlock
  | KPICard
  | TableBlock
  | ChartBlock
  | InsightBlock
  | ComparisonBlock
  | ForecastBlock
  | RiskMatrixBlock
  | ExecutiveSummaryBlock

export interface TOCEntry {
  block_id: string
  title: string
  block_type: string
  order: number
}

export interface ReportMetadata {
  total_tokens_used: number
  processing_time_seconds: number
  departments_involved: string[]
  files_analyzed: string[]
  llm_provider: string
  model_name: string
  quality_score: number
  partial_blocks_count: number
}

export interface Report {
  report_id: string
  title: string
  created_at: string
  query: string
  blocks: ReportBlock[]
  toc: TOCEntry[]
  metadata: ReportMetadata
}

export interface SSEEvent {
  type: 'block_approved' | 'task_status' | 'session_done' | 'error'
  [key: string]: unknown
}

export interface TaskStatusEvent {
  type: 'task_status'
  task_id: string
  department: string
  status: string
  retries?: number
  tokens_used?: number
}

export interface UploadedFile {
  file_id: string
  filename: string
  file_type: string
  size_bytes: number
  status: string
}

File: backend/app/schemas/blocks.py
from __future__ import annotations
from datetime import datetime
from typing import Literal, Union, Annotated, Any
from pydantic import BaseModel, Field
from enum import Enum
import uuid


class BlockStatus(str, Enum):
    COMPLETE = "complete"
    PARTIAL = "partial"
    ERROR = "error"


class ReportBlockBase(BaseModel):
    block_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    block_type: str
    title: str
    order: int = 0
    status: BlockStatus = BlockStatus.COMPLETE
    warnings: list[str] = []
    source_task_ids: list[str] = []
    created_by_dept: str = ""
    created_at: datetime = Field(default_factory=datetime.utcnow)


# ─── Text ────────────────────────────────────────────────────────────────────

class TextBlock(ReportBlockBase):
    block_type: Literal["text"] = "text"
    content: str
    style: Literal["body", "heading", "callout", "quote"] = "body"
    language: str = "ru"


# ─── KPI Card ────────────────────────────────────────────────────────────────

class KPICard(ReportBlockBase):
    block_type: Literal["kpi_card"] = "kpi_card"
    metric_name: str
    value: float
    unit: str = ""
    delta: float | None = None
    delta_pct: float | None = None
    trend: Literal["up", "down", "stable"] | None = None
    is_positive_trend: bool | None = None
    benchmark: float | None = None
    verification_status: Literal["verified", "unverified"] = "verified"
    source_reference: str = ""


# ─── Table ───────────────────────────────────────────────────────────────────

class TableColumn(BaseModel):
    key: str
    label: str
    dtype: Literal["number", "string", "date", "percent", "currency"] = "string"
    unit: str | None = None
    sortable: bool = True
    highlight_rule: dict[str, Any] | None = None


class TableBlock(ReportBlockBase):
    block_type: Literal["table"] = "table"
    columns: list[TableColumn]
    rows: list[dict[str, Any]]
    totals_row: dict[str, Any] | None = None
    pagination: bool = True
    exportable: bool = True
    source_file_id: str = ""
    source_location: str = ""


# ─── Chart ───────────────────────────────────────────────────────────────────

class ChartType(str, Enum):
    BAR = "bar"
    LINE = "line"
    AREA = "area"
    PIE = "pie"
    SCATTER = "scatter"
    HEATMAP = "heatmap"
    HISTOGRAM = "histogram"
    BOX = "box"
    WATERFALL = "waterfall"
    TREEMAP = "treemap"
    FUNNEL = "funnel"
    RADAR = "radar"


class ChartBlock(ReportBlockBase):
    block_type: Literal["chart"] = "chart"
    chart_type: ChartType
    vega_lite_spec: dict[str, Any]
    caption: str | None = None
    data_source_description: str = ""
    interactive: bool = True


# ─── Insight ─────────────────────────────────────────────────────────────────

class InsightBlock(ReportBlockBase):
    block_type: Literal["insight"] = "insight"
    insight_type: Literal["finding", "anomaly", "recommendation", "risk", "opportunity"]
    severity: Literal["info", "warning", "critical"] = "info"
    headline: str
    explanation: str
    supporting_data: list[str] = []
    confidence: float = 0.8


# ─── Comparison ──────────────────────────────────────────────────────────────

class ComparisonItem(BaseModel):
    label: str
    metrics: dict[str, float | str]


class ComparisonBlock(ReportBlockBase):
    block_type: Literal["comparison"] = "comparison"
    items: list[ComparisonItem]
    dimensions: list[str]
    winner: str | None = None
    analysis: str = ""


# ─── Forecast ────────────────────────────────────────────────────────────────

class TimePoint(BaseModel):
    period: str
    value: float


class ForecastPoint(BaseModel):
    period: str
    value: float
    lower_bound: float
    upper_bound: float


class ForecastBlock(ReportBlockBase):
    block_type: Literal["forecast"] = "forecast"
    metric_name: str
    historical: list[TimePoint]
    forecast: list[ForecastPoint]
    method: str = "auto"
    confidence_interval: float = 0.95
    model_accuracy: dict[str, float] = {}


# ─── Risk Matrix ─────────────────────────────────────────────────────────────

class RiskItem(BaseModel):
    name: str
    probability: Literal["low", "medium", "high"]
    impact: Literal["low", "medium", "high"]
    description: str
    mitigation: str | None = None


class RiskMatrixBlock(ReportBlockBase):
    block_type: Literal["risk_matrix"] = "risk_matrix"
    risks: list[RiskItem]
    overall_risk_level: Literal["low", "medium", "high", "critical"]


# ─── Executive Summary ────────────────────────────────────────────────────────

class ExecutiveSummaryBlock(ReportBlockBase):
    block_type: Literal["executive_summary"] = "executive_summary"
    key_findings: list[str]
    recommendations: list[str]
    overall_conclusion: str
    report_quality_score: float = 0.8


# ─── Discriminated Union ─────────────────────────────────────────────────────

ReportBlock = Annotated[
    Union[
        TextBlock,
        KPICard,
        TableBlock,
        ChartBlock,
        InsightBlock,
        ComparisonBlock,
        ForecastBlock,
        RiskMatrixBlock,
        ExecutiveSummaryBlock,
    ],
    Field(discriminator="block_type"),
]

BLOCK_TYPES = {
    "text": TextBlock,
    "kpi_card": KPICard,
    "table": TableBlock,
    "chart": ChartBlock,
    "insight": InsightBlock,
    "comparison": ComparisonBlock,
    "forecast": ForecastBlock,
    "risk_matrix": RiskMatrixBlock,
    "executive_summary": ExecutiveSummaryBlock,
}
